gtsam/hybrid/HybridFactorGraph.cpp

```diff
@@ -23,190 +23,33 @@
-  nonlinearGraph_.print(prefix + "NonlinearFactorGraph", keyFormatter);
-  discreteGraph_.print(prefix + "DiscreteFactorGraph", keyFormatter);
-  dcGraph_.print(prefix + "DCFactorGraph", keyFormatter);
-  gaussianGraph_.print(prefix + "GaussianGraph", keyFormatter);
-}
-
-HybridFactorGraph HybridFactorGraph::linearize(
-    const Values& continuousValues) const {
-  // linearize the continuous factors
-  auto gaussianFactorGraph = nonlinearGraph_.linearize(continuousValues);
-
-  // linearize the DCFactors
-  DCFactorGraph linearized_DC_factors;
-  for (auto&& dcFactor : dcGraph_) {
-    // If dcFactor is a DCGaussianMixtureFactor, we don't linearize.
-    if (boost::dynamic_pointer_cast<DCGaussianMixtureFactor>(dcFactor)) {
-      linearized_DC_factors.push_back(dcFactor);
-    } else {
-      auto linearizedDCFactor = dcFactor->linearize(continuousValues);
-      linearized_DC_factors.push_back(linearizedDCFactor);
-    }
-  }
-
-  // Add the original factors from the gaussian factor graph
-  for (auto&& gaussianFactor : gaussianGraph()) {
-    gaussianFactorGraph->push_back(gaussianFactor);
-  }
-
-  // Construct new linearized HybridFactorGraph
-  return HybridFactorGraph({}, discreteGraph_, linearized_DC_factors,
-                           *gaussianFactorGraph);
-}
-
-bool HybridFactorGraph::equals(const HybridFactorGraph& other,
-                               double tol) const {
+template <typename FG>
+bool HybridFactorGraph<FG>::equals(const HybridFactorGraph<FG>& other,
+                                   double tol) const {
   return Base::equals(other, tol) &&
-         nonlinearGraph_.equals(other.nonlinearGraph_, tol) &&
          discreteGraph_.equals(other.discreteGraph_, tol) &&
          dcGraph_.equals(other.dcGraph_, tol) &&
-         gaussianGraph_.equals(other.gaussianGraph_, tol);
-}
-
-void HybridFactorGraph::clear() {
-  nonlinearGraph_.resize(0);
-  discreteGraph_.resize(0);
-  dcGraph_.resize(0);
-  gaussianGraph_.resize(0);
-}
-
-DiscreteKeys HybridFactorGraph::discreteKeys() const {
-  DiscreteKeys result;
-  // Discrete keys from the discrete graph.
-  result = discreteGraph_.discreteKeys();
-  // Discrete keys from the DC factor graph.
-  auto dcKeys = dcGraph_.discreteKeys();
-  for (auto&& key : dcKeys) {
-    // Only insert unique keys
-    if (std::find(result.begin(), result.end(), key) == result.end()) {
-      result.push_back(key);
-    }
-  }
-  return result;
+         factorGraph_.equals(other.factorGraph_, tol);
 }
 
-/// Define adding a GaussianFactor to a sum.
 using Sum = DCGaussianMixtureFactor::Sum;
-static Sum& operator+=(Sum& sum, const GaussianFactor::shared_ptr& factor) {
-  using Y = GaussianFactorGraph;
-  auto add = [&factor](const Y& graph) {
-    auto result = graph;
-    result.push_back(factor);
-    return result;
-  };
-  sum = sum.apply(add);
-  return sum;
-}
 
-Sum HybridFactorGraph::sum() const {
+template <typename FG>
+Sum HybridFactorGraph<FG>::sum() const {
```

Comment:
template should be in header. Or just specialize it on GaussianFG and move there.